### DISCUSSION TITLE
Which parameters are independent for loading a microstructure?

### DISCUSSION THREAD

**Discussion by CsatiZoltan** [neperfepx/FEPX#21] [Q&A] [answered]
Hi,

In the configuration file, I need to give `target_strain`, `read_bcs_from_file` and `strain_rate`. I already ran a simulation successfully in which I imposed
```
...
target_strain 0.005 4 print_data
target_strain 0.010 4 print_data
target_strain 0.015 2 print_data
target_strain 0.020 2 print_data
...
read_bcs_from_file
...
strain_rate 1e-2
```
As the manual writes: *"The velocities should be prescribed relative to the mesh dimensions and time-step size in order to produce expected strain rates."*, I took into account the size of the specimen (0.15 mm), so I prescribed v=0.0015 mm/s at the selected nodes so as to replicate uniaxial tension. Everything went fine.

This time, I use externally defined boundary conditions (velocities) that come from DIC measurements. From our measurements, the target strain, the strain rate, and the nodal velocities are all given by the experiment. This leaves no free parameter to set. Hence, my question is: which quantities are fixed in FEPX and which ones are computed? In other words, how should I set the parameters to reconstruct the experimental setting? I tend to say that the velocities are definitely fixed by the experiment as we will perform an inverse analysis, in which velocities are Dirichlet boundary conditions.

Thank you,
Zoltan

**Comment by acmelab-ua** [comment]
@CsatiZoltan,

I'm a bit confused as to what you're doing. Are you attempting to apply velocities to all nodes within a polycrystal? Your domain is 3 dimensional, so how do you know the velocities at every point?

Boundary conditions are just that: essential velocities that are prescribed at the boundary of the domain. All other nodal velocities should be free to evolve according to the constitutive laws and evolution equations of the model. The boundary conditions should best match the macroscopic loading of your experiment.

So, let's assume you are doing the latter: you are applying boundary conditions on two parallel surfaces to mimic a uniaxial tensile experiment. Firstly, why do you need to apply boundary conditions from a file, when you could likely use one of the uniaxial boundary conditions to handle this?

To answer more directly: the strain rate and the nodal velocities are coupled. When you select built-in boundary conditions, you prescribe a strain rate. The nodal velocities are then calculated to adhere to the set strain rate (based on the size of the domain). When you supply your own boundary conditions, which contain velocities on a certain control surface, your strain rate should match that which would be calculated given the velocities in your boundary condition file.

I do not understand what you mean by there is no free parameter to set. Velocities and strain rates are coupled, and strain targets are set such that appropriate time steps are taken in the simulation to reach said targets.

Can you please better describe (perhaps with a sketch) what boundary conditions you are trying to apply?

**Reply by CsatiZoltan** [reply]
@acmelab-ua 
I try to explain my problem better.
Part of the original microstructure can be seen here.
![microstructure](https://user-images.githubusercontent.com/12742890/121998649-7707c900-cdac-11eb-97eb-6f4b1fe3e1fc.png)
At every pixel, the displacement and the velocity fields are available from the experiment (uniaxial tensile test with DIC measurement). As the machine recorded the macroscopic strain and strain rates, they are also available.
Since most crystal plasticity frameworks -- including FEPX -- are 3D, we made the following modelling choices.
1. Assume that the grains are columnar (see the first picture in the figure below).
2. Assume that the field values (displacement, velocity) are uniform in the third dimension.
   In the second picture of the figure, you can see that I defined the top face (in transparent red) and the four side faces (let us call them *walls*). The top face corresponds to the original two-dimensional microstructure.

The generated mesh can be seen in the third picture of the figure, where face and node sets are available so that I can define the boundary conditions later.
![3Dmodel](https://user-images.githubusercontent.com/12742890/122000445-1d54ce00-cdaf-11eb-8520-03016f55093d.png)

Our initial goal is to have an understanding on how boundary conditions influence the numerical solution, and to obtain a model that matches the experimentally obtained result. So we want to prescribe the boundary conditions such that the obtained solution (restricted onto the top face) is close to the experimental field.
Concerning the boundary conditions, I want to prescribe the following:
1. "Symmetry" boundary condition on the top face, i.e. the normal velocity is zero, while the in-plane components are not prescribed.
2. The boundary conditions on the walls come from the experiment. That is, the values on the boundary of the original square are applied on the walls, assuming uniform distribution along the extruded dimension (axis z).

Based on the explanation above, I can now answer to your comments.

> Are you attempting to apply velocities to all nodes within a polycrystal?

No, only at the nodes on the top face and on the four walls.

> Your domain is 3 dimensional, so how do you know the velocities at every point?

I know the velocities at every point (every pixel, and hence every node by interpolation) of the original, two-dimensional domain. Making the assumption that the velocity distribution is uniform in the third dimension (i.e. along the extrusion), we know the velocities at each point in the three-dimensional domain as well.

> Boundary conditions are just that: essential velocities that are prescribed at the boundary of the domain. All other nodal velocities should be free to evolve according to the constitutive laws and evolution equations of the model.

Yes, and that is what we do: prescribe boundary conditions on 5 boundary faces of the 3D geometry. I did not mention the extrusion of the geometry from 2D to 3D, and I think that is what confused you.

> The boundary conditions should best match the macroscopic loading of your experiment.

Thanks to the full-field measurement, we have access to the velocities at each point, including the boundaries as well. Therefore, we do not need to manually set the boundary velocities according to the macroscopic loading. It also sheds light on the fact that our experiment left no free parameters: the boundary conditions are fully known, as well as the macroscopic loading conditions (strain and strain rate). That is why I asked how FEPX works.

> So, let's assume you are doing the latter: you are applying boundary conditions on two parallel surfaces to mimic a uniaxial tensile experiment. Firstly, why do you need to apply boundary conditions from a file, when you could likely use one of the uniaxial boundary conditions to handle this?

Now you know that we have rich data from the DIC, so we want to make use of that. That is why I do not simply apply velocities according to uniaxial boundary conditions (as in example 1, shipped with the FEPX code). Rather, I prescribe the experimentally obtained boundary velocities. For that, I need to give them in an external file. The technical details on how to give them in the `simulation.bcs` file in not a problem for me, I already coded it.

> When you supply your own boundary conditions, which contain velocities on a certain control surface, your strain rate should match that which would be calculated given the velocities in your boundary condition file.

When velocities are uniform on the loaded face, one can indeed provide a strain rate that gives rise to the prescribed velocities. However, I prescribe non-uniform velocity boundary conditions in the file. Then how to set the strain rate?

> I do not understand what you mean by there is no free parameter to set. Velocities and strain rates are coupled, and strain targets are set such that appropriate time steps are taken in the simulation to reach said targets.

In my case, velocities on the boundary are known from the DIC, the macroscopic strain and strain rate are also known (recorded by the testing device). That is what I meant that there is no free parameter to set. In case of the uniaxial boundary condition you explained in your answer, you set the strain rate so that it matches the *uniform* velocities you prescribe. This is not the case in our setup. First, you cannot give a strain rate that will give rise to the non-uniform velocity boundary conditions. Second, you cannot give the strain rate arbitrarily, as it was recorded by the device, so it is given. This is why I asked what parameters are fixed in FEPX, and which ones are computed, so that I can give the appropriate values in my `simulation.config` file.

**Reply by acmelab-ua** [reply]
@CsatiZoltan: I'm still confused as to what you are attempting to do. Please better describe your boundary conditions in a structured way. You have 6 faces, x0, x1, y0, y1, z0, z1, corresponding to the faces where x, y, and z are minimum and maximum. You can apply velocities in any direction, x, y, and z on any of these 6 faces. What are your boundary conditions?

Example:

"Grip" boundary conditions for uniaxial extension in the "z" direction. Grip boundary conditions assume a control surface and a fixed surface move apart from one another (i.e., in the direction of the surface normal), and the area is held fixed. The boundary conditions thus are:

x0, x1, y0, y1: No BCs.
z0: x, y, and z velocities held fixed at 0.
z1: x and y velocities held fixed at 0, z velocity non-zero, scaled to provide desired strain rates.

**Reply by CsatiZoltan** [reply]
@acmelab-ua My boundary conditions are the following:
- x0, x1: z velocity 0, x and y velocities change with respect to the y coordinate only and their values match that of the experiment
- y0, y1: z velocity 0, x and y velocities change with respect to the x coordinate only and their values match that of the experiment
- z1: z velocity 0
- z0: no BCs

**Reply by rquey** [reply]
@CsatiZoltan, I think you have to set the `strain_rate` from the _average y velocities_ on the _control surface_ (`x1`, and the length of the sample, along _x_). There is only one control surface/direction. As for the other velocities, the externally loaded velocities suffice (they are not related to `target_strain` and `strain_rate`).

@mkasemer, does it make sense?

**Reply by CsatiZoltan** [reply]
@rquey 
> I think you have to set the strain_rate from the average y velocities on the control surface (x1, and the length of the sample, along x).

That's what I thought too, to somehow trace this configuration back to a uniaxial loading case, but why the average *y* and not the average *x* velocities?

> As for the other velocities, the externally loaded velocities suffice (they are not related to `target_strain` and `strain_rate`)

Thank you, this is a useful piece of information I did not know.

**Reply by rquey** [reply]
> > I think you have to set the strain_rate from the average y velocities on the control surface (x1, and the length of the sample, along x).
> 
> That's what I thought too, to somehow trace this configuration back to a uniaxial loading case, but why the average _y_ and not the average _x_ velocities?

If they don't average to zero, I think it would not be impossible to use the _x_ velocities instead.

Let's wait and hear back from @mkasemer.

**EDIT** I've already done this, except that the imposed velocities were uniform on the surfaces and only normal.

**Reply by acmelab-ua** [reply]
@CsatiZoltan: so you are applying biaxial loading, then? Just to be clear? If truly uniaxial, you should be considering only one surface with non-zero velocities.

@rquey is right in his assessment that even for triaxial loading, there is only one control surface.

But, you still need to designate a strain rate. The internal calculation of time steps is based on the target strains and the strain rate.

**Reply by CsatiZoltan** [reply]
@acmelab-ua: The experiment was a uniaxial test, but locally the cropped part of the microstructure (300 x 300 pixel, see the first image I attached above) experiences heterogeneous velocity field. Therefore, if we apply such loading on the cropped part of the microstructure, that will be under general loading condition, not under *uniaxial* or *multiaxial* as far as these terms are defined in Section 2.1.2 of the documentation.

**EDIT**: @rquey : 
> I've already done this, except that the imposed velocities were uniform on the surfaces and only normal.

Yes, that's the difference between our case and what you have done before. I guess there is no confusion any more.

**Comment by mkasemer** [comment, marked as answer]
Yes, so I'm fairly sure that @rquey is right regarding the imposition of your strain rate (the average across the surface that you deem the control surface).

Though, I'm still not sure about how appropriate this setup is. I'm having a hard time seeing the significance in the polycrystal design and the applied "boundary conditions". It's essentially a pseudo-2D setup, and you already know all of the nodal velocities (and thus displacements). Why do you need complex simulations to calculate anything when it can be directly calculated (if you already know the velocity field)?

**Reply by CsatiZoltan** [reply]
@mkasemer 

> Why do you need complex simulations to calculate anything when it can be directly calculated (if you already know the velocity field)?

Indeed, the microstructure is already given (revealed by electron microscopy), and the displacement and velocity fields have already been measured (by DIC). However, this was an extremely daunting task. Therefore, simulations can help in the future to
- partially replace measurements
- serve as ground truth for simplified (numerical) models (crystal plasticity is computationally demanding),

Performing all the necessary measurements for each microstructure in the industry is infeasible. That is why we want to establish a simulation framework that is tuned to capture the real-world behaviour.

**Reply by rquey** [reply]
For having seen this already, I think @CsatiZoltan is trying to compare an experiment where all displacements are known on the surface (DIC experiment) with a simulation. In the simulation, an internal region of the surface of the experimental sample is considered, and to avoid any errors on the boundary displacements (errors you would get if the displacement was imposed uniform, since it is not), the experimental displacements are imposed. The simulated displacements are the boundary will indeed necessarily be correct, but the full displacement field (internal nodes of the simulated region) will match the experimental one only if the crystal constitutive law used in the simulation is correct (hence the comparison between experiment and simulation). It is one possible choice to compare experiment and simulation.

**Reply by mkasemer** [reply]
@CsatiZoltan: This is precisely why standard boundary conditions are important: you cannot measure velocity fields via SEM/DIC for every microstructure. The simulations allow you to best replicate general experimental loading through standard BCs, and to simulate the deformation fields based on the constitutive models and state evolution equations. Using DIC displacement measurements (from which velocities are integrated) is appropriate for only one microstructure, and again requires no simulations: if you know the velocity field, the stress distributions may be calculated directly.

**Reply by mkasemer** [reply]
> For having seen this already, I think @CsatiZoltan is trying to compare an experiment where all displacements are known on the surface (DIC experiment) with a simulation. In the simulation, an internal region of the surface of the experimental sample is considered, and to avoid any errors on the boundary displacements (errors you would get if the displacement was imposed uniform, since they are not), the experimental displacements are imposed. The simulated displacements are the boundary will indeed necessarily be correct, but the full displacement field (internal nodes of the simulated region) will match the experimental one only if the crystal constitutive law used in the simulation is correct (hence the comparison between experiment and simulation). It is one possible choice to compare experiment and simulation.

_Ah_ - I now better understand the setup. Thank you @rquey.

My comment here is this: even knowing the precise displacements (and thus velocities) at the "boundary" of this subset region, you are still subject to effects of the boundary conditions. My suggestion - the same as if you were to be performing a simulation on a randomly generated polycrystal - would be to simulate a larger region with "standard" BCs, with sufficient distance between the region you are interested in and the applied boundary conditions. In this way, you can mitigate the effects that the essential boundary conditions have on the region you are analyzing - if the grains you are interested in are sufficiently far from the boundary, then the primary effect on their deformation response is going to be from their neighboring grains (i.e., the constitutive model) rather than the applied BCs. 

I'm afraid that even with a relatively high degree of confidence in the displacements at the edge of the region you are interested in, there could be influences due to these BCs that lead to error.

**Reply by CsatiZoltan** [reply]
@rquey This is precisely the goal!
@mkasemer Romain Quey nicely summarized the procedure. I would add that once we have detailed experimental data (and we do have detailed measurements, also the grain orientations by EBSD), we can fine-tune the numerical setting such that it will remain predictive for further microstructures as well, in which case no detailed measurements will be at our disposal.

**Reply by CsatiZoltan** [reply]
@mkasemer 
> simulate a larger region with "standard" BCs

True, but that has two drawbacks. First, simulating the whole region will be very costly as our mesh consists of many elements (as opposed to generated microstructures in which high-quality and relatively coarse meshes can be obtained, our grains are highly complex, requiring a fine mesh). Second, the displacement field still needs to be considered in a smaller region than what corresponds to the test sample due to uncertainties in the DIC measurements. This means, that even if a large region was considered, the displacement field on its boundary is far from being uniaxial, as we already noticed that.

**Reply by mkasemer** [reply]
@CsatiZoltan: when including "dummy" (or buffer) regions, there are techniques to reduce mesh size, specifically by selectively coarsening grains in which you have little interest in their overall response. 

I should also note that (for regions in which you _are_ interested in) you should strive to reduce selective refinement of your mesh (i.e., regions with significantly reduced element size compared to the average) for reasons both practical (over-refined regions may lead to premature failure of a simulation due to element inversion) and numerical (over-refined regions will lead to biased results).

These two things should can be considered both to contribute to a reduction in mesh size, as well as lend greater confidence in the intra-grain deformation predictions.

The second point is fair, though you can test to see what effects the BCs have on the subset of interest (i.e., whether the response is sensitive to changes in the standard BCs or not). At a sufficient distance, it should not (unless there is a significant change in loading, i.e. a large difference in biaxiality ratios).

**Reply by CsatiZoltan** [reply]
@mkasemer Thank you for the remarks on mesh refinement/coarsening.
> there are techniques to reduce mesh size

Could you provide me some references or some keywords? For mesh generation, I use the Netgen plugin of Salome, which generates high-quality but overly fine meshes. For mesh decimation/coarsening, I had some moderately successful attempts with [Mmg](https://www.mmgtools.org/).

I will try standard boundary conditions and the one we have discussed so far (i.e. prescribing the measured velocities). Concerning the load control in the latter case, I will stay with the uniaxial assumption in FEPX and setting the strain rate as Romain Quey proposed. As for the target strain, I will proceed as the experiment dictates in order to give a reasonable value. I hope this is what you think of too.

**Reply by mkasemer** [reply]
Concerning meshing techniques, @rquey may be a better source for suggestions (he is certainly the expert among the two of us).

If you must use the "subset BCs" (for a lack of better term), I think the way that @rquey described should work. One of the faces must be the control face, and the strain rate and strain targets must be associated with said control face. Otherwise it is ambiguous. Note that you will have to provide FEPX with a control surface and an extension direction. Please let us know if you have any other issues!

Overall, strain rate should be redundant if the velocities have the correct magnitude... this brings up the point that either 1) strain rate as given by the user should be enforced (regardless of average velocity magnitude in the extension direction), or 2) strain rates should be internally calculated based on the average velocity magnitude in the extension direction... otherwise, if the user supplied a strain rate that does not "agree" with the velocities, the time steps will be calculated and the simulated strains will be different than the expected targets... something to mull for next release.

**Reply by CsatiZoltan** [reply]
@mkasemer 

> Note that you will have to provide FEPX with a control surface and an extension direction.

I guess you think of giving in the configuration file the following (assuming that the direction of loading is Z):
```
loading_direction Z
loading_face Z_MAX
```
If this is what you think of, then I have a remark. Namely, when I first tested user-defined BCs, I wanted to check whether I can get back the usual uniaxial loading case (i.e. the example "1_uniaxial" in the FEPX program code). It turned out that I didn't need to provide the two lines above. In other words, the settings
```
boundary_conditions uniaxial_grip
loading_direction Z
loading_face Z_MAX
strain_rate 1e-2
```
in "1_uniaxial" gave the same result as
```
read_bcs_from_file
strain_rate 1e-2
```
with the velocities properly defined in my `simulation.bcs` file. Thus, giving the control surface and the extension direction seems to be irrelevant when velocities are provided at each node -- which makes sense.

> something to mull for next release

I would opt for strategy number 2. If you discard the given velocities, the user cannot handle cases like mine (when the prescribed velocity field is non-uniform).

**Reply by mkasemer** [reply]
To clarify the last point: rather than _discard_ the velocities, they would be normalized such that their average velocity in the extension direction was 1. Their directions and relative magnitudes would be retained.

I suspect for the other portion that the default loading direction and loading face are those that are prescribed there. If your direction or face differ, they should likely be specified.

**Reply by CsatiZoltan** [reply]
All right. In the upcoming days, I will experiment with the boundary conditions as we discussed. I will come back when I have results or further questions. Thanks for the help, all of you.

**Reply by CsatiZoltan** [reply]
@mkasemer I ran some simulations and my observations are below. I set the loading surfaces and specified the type of loading, where `strain_rate` is set to be the mean velocity on the `X1` face divided by the length in that direction (distance between the faces `X0` and `X1`). The boundary conditions were such that the normal velocity on the `X0` and `Z1` faces was 0, the normal velocity on the `X1` face was set to be uniform, and the other two velocity components on this face were set to 0. The goal of this configuration was to test the effect of the strain rate.
```
def_control_by uniaxial_strain_target

number_of_strain_steps 1
target_strain 0.005 4 print_data

read_bcs_from_file

loading_direction X
loading_face X_MAX

strain_rate 0.00043166438            
```

> Overall, strain rate should be redundant if the velocities have the correct magnitude...

The strain rate does have an effect on the results, although I set custom boundary conditions. The strain rate controls the time steps, as I see from the output, i.e. `strain_rate` * *end_time* = `target_strain`. On the other hand, using
```
strain_rate 1e-2
```
gave *end_time* = 0.5, which is logical, but the $E_{11}$ strain distribution and their magnitude also changed. With larger strain rate, i.e. with 1e2, the strain values became smaller. It seems to me that the smaller time steps caused FEPX not to reach the strain target specified. Can you explain this? I want to understand first this simplified configuration before going for the one you named "subset BCs", i.e. where the velocity is specified and non-uniform on `X0`, `X1`, `Y0`, `Y1`.

**Reply by acmelab-ua** [reply]
Yes, it was incorrect for me to state redundant. As stated previously, the internal time steps are calculated based on the strain rate and the target strains.

What you described makes sense. You must manipulate your above equation to see what is done internally:

`time = target_strain / strain_rate`

So, you prescribe the target strain and the strain rate. FEPX then calculates the time steps. If your strain rate is artificially high (i.e., it does not match the expected strain rate based on the velocities prescribed in your boundary conditions`, than the time step is going to be artificially low for the desired target strain. Your actual strain, then, will undershoot: if you have a small time step, it will reach a smaller strain than expected.

Does this make sense? It is, again, an issue with regard to the custom BCs (rarely used!) and must be better documented to remove the possibility of such error. It is difficult to apply a general fix for this: it is easy if we assume a cubic domain with constant velocities on a control face - FEPX could calculate the strain rate very easily internally, and the user would not have to specify it. However, it becomes an issue with specific domains, loadings, etc...

**Reply by CsatiZoltan** [reply]
@acmelab-ua 
> Does this make sense?

Yes, that is what I thought as well. Thank you for assuring me.

> It is difficult to apply a general fix for this

Trial and error would suffice if I only had to consider one configuration. However, we have a series of measurements, so the idea is the following. Start with the first data set and compute the response with FEPX. Then perform a restart analysis to continue the loading of the body (now with different boundary conditions!). Since we have 382 steps to take, it must be automated. If nothing else works, I may try to write a predictor-corrector method, i.e.
1. For data step  *i*, continue the restart analysis from data step *i-1*. Give the target strain for step *i*, being available from the load-displacement curve of the experiment.
2. Compute the strain rate based on the averaging strategy described above.
3. Perform the simulation with that candidate strain rate value (predictor).
4. If the target strain is not reached (or surpassed), go to step 2 (corrector).
5. Once the target stain for step *i* has been approximated to a given tolerance, accept the result and move to step *i+1*.

This will surely require tons of computations.

----
I already experienced, and you have just explained above, that when the specified strain rate is too large, the time step is too low to reach the given strain target. What happens when the strain rate is too small, and hence the time step is too large? Will FEPX stop when the target strain (displacement of the control surface divided by the initial length of the body) is reached, or it will overshoot? Knowing this information would help me in designing the predictor-corrector algorithm.

**Reply by acmelab-ua** [reply]
@CsatiZoltan: Do you mean that you will change the boundary conditions after every single load step, based on experimental data? If this is the intent, you will likely have to edit the code to handle this very specific case...

To be clear: when we restart a simulation, we load in the velocity field from the previous load step, saved to the restart output files. The simulation then continues with this velocity field and marches forward in time based on the additional load step(s). I'm not sure that it will re-consider the boundary conditions, and you may have to edit the code to allow for this custom functionality (I'm unsure without inspecting closely).

Then, your steps as described would be correct. I'm not sure you necessarily need step 4... if your strain rate is calculated correctly, and your target strain set, you will calculate the necessary time step correctly. Everything will then be in-sync with one another. The `velocitiy * time` on average will lead to displacements which will give you (again, on average) the macroscopic strain you desire. The computation for this work flow should not be terribly difficult - you will simply have to calculate the strain-rate based on the average. Everything else would, presumably, be able to be handled by FEPX.

You can check to see if FEPX reads in new boundary conditions properly upon restart by printing the velocity field at only the points BCs are set, i.e., `VELOCITY(BCS)`, _after_ the restart file has been parsed. I'm afraid that you will read in your custom BCs and velocities, but the velocity field will then be read-in, overwriting the essential velocities that were set when the BCs were read-in, and the simulation will continue with the _previous_ essential velocities.

**Reply by CsatiZoltan** [reply]
> Do you mean that you will change the boundary conditions after every single load step, based on experimental data?

Yes, that's exactly the goal. By load step, I do not mean strain or load increments, or multiple strain targets in the configuration file, rather a new analysis. E.g. the `4_restart` example shipped with the FEPX code defines two analysis steps, `cycle1` and `cycle`. In my case, I will have not 2, but 382.

> I'm not sure you necessarily need step 4... if your strain rate is calculated correctly, and your target strain set, you will calculate the necessary time step correctly. Everything will then be in-sync with one another. The velocitiy * time on average will lead to displacements which will give you (again, on average)

For uniaxial loading (i.e. with boundary conditions defined in the FEPX manual), they will be in sync, but for the specific boundary conditions I apply, it may not be the case I think. Because for my non-proportional loading, there is no longer linear relationship. Anyway, I will try that.

> I'm afraid that you will read in your custom BCs and velocities, but the velocity field will then be read-in, overwriting the essential velocities that were set when the BCs were read-in, and the simulation will continue with the previous essential velocities.

Hmmm... That is a problem. On the other hand, if I am not mistaken, FEPX performs the restart analysis using file input-output, and not in-memory. So I guess, I can programmatically modify the restart file with the new velocity BCs, without having to dig into the source code of FEPX and recompile it (that would be a pity, as I cannot program in Fortran).
In the meantime, I checked `4_restart`, but that will not help me as in that example the boundary conditions stay the same for the two load steps:
```
boundary_conditions triaxial
loading_direction Z
```

**Reply by acmelab-ua** [reply]
The BCs are generally static (or iterated upon under very constrained and defined circumstances), so this would be a custom case that is not covered by the examples.

Yes, restart is performed using restart files that are written in binary format. The restart files could, ostensibly, be edited to change the essential velocities. I think it would be easier to handle in-code. Please email me at mkasemer@eng.ua.edu if you'd like to discuss more about this.
